internal/internal_workflow_client_test.go

```diff
@@ -1756,6 +1768,704 @@ func (s *workflowClientTestSuite) TestTerminateWorkflow() {
+			},
+
+			rpcResponse: &shared.QueryWorkflowResponse{QueryResult: []byte("\"result\"")},
+			rpcError:    nil,
+			responseValidator: func(resp *QueryWorkflowWithOptionsResponse, err error) {
+				s.Require().Nil(err)
+				s.Nil(resp.QueryRejected)
+
+				var res string
+				s.NoError(resp.QueryResult.Get(&res))
+				s.Equal("result", res)
+			},
+		},
+		{
+			name:      "success with arguments",
+			queryArgs: []interface{}{"arg1", "arg2"},
+			expectRPC: true,
+			requestValidator: func(req *shared.QueryWorkflowRequest) {
+				s.Equal("\"arg1\"\n\"arg2\"\n", string(req.GetQuery().GetQueryArgs()))
+			},
+
+			rpcResponse: &shared.QueryWorkflowResponse{QueryResult: []byte("\"result\"")},
+			rpcError:    nil,
+			responseValidator: func(resp *QueryWorkflowWithOptionsResponse, err error) {
+				s.Require().Nil(err)
+				s.Nil(resp.QueryRejected)
+
+				var res string
+				s.NoError(resp.QueryResult.Get(&res))
+				s.Equal("result", res)
+			},
+		},
+		{
+			name:      "failed to encode arguments",
+			queryArgs: []interface{}{make(chan int)}, // you can't marshal this object to JSON
+			expectRPC: false,
+
+			responseValidator: func(resp *QueryWorkflowWithOptionsResponse, err error) {
+				s.ErrorContains(err, "unable to encode")
+				s.Nil(resp)
+			},
+		},
+		{
+			name:             "RPC fails",
+			queryArgs:        nil,
+			expectRPC:        true,
+			requestValidator: func(req *shared.QueryWorkflowRequest) {},
+
+			rpcResponse: nil,
+			rpcError:    &shared.AccessDeniedError{},
+			responseValidator: func(resp *QueryWorkflowWithOptionsResponse, err error) {
+				s.Equal(&shared.AccessDeniedError{}, err)
+				s.Nil(resp)
+			},
+		},
+		{
+			name:             "query rejected",
+			queryArgs:        nil,
+			expectRPC:        true,
+			requestValidator: func(req *shared.QueryWorkflowRequest) {},
+
+			rpcResponse: &shared.QueryWorkflowResponse{
+				QueryRejected: &shared.QueryRejected{
+					CloseStatus: shared.WorkflowExecutionCloseStatusTerminated.Ptr(),
+				},
+			},
+			rpcError: nil,
+			responseValidator: func(resp *QueryWorkflowWithOptionsResponse, err error) {
+				s.Require().Nil(err)
+				s.Nil(resp.QueryResult, "should be nil when query rejected")
+
+				s.Require().NotNil(resp.QueryRejected)
+				s.Equal(shared.WorkflowExecutionCloseStatusTerminated, resp.QueryRejected.GetCloseStatus())
+			},
+		},
+	}
+
+	for _, tt := range testcases {
+		s.Run(tt.name, func() {
+			if tt.expectRPC {
+				s.service.EXPECT().
+					QueryWorkflow(gomock.Any(), gomock.Any(), gomock.Any()).
+					Do(func(_ context.Context, req *shared.QueryWorkflowRequest, _ ...yarpc.CallOption) {
+						// do common validation for some fields
+						s.Equal(domain, req.GetDomain())
+						s.Equal(workflowID, req.GetExecution().GetWorkflowId())
+						s.Equal(runID, req.GetExecution().GetRunId())
+						s.Equal(queryType, req.GetQuery().GetQueryType())
+						// ... and custom validation for the testcase
+						tt.requestValidator(req)
```

Comment:
nit: you can check for nil and avoid this call if it is not provided.
nit: you can also move common checks to a function and explicitly provide it as a default value.

side nit: each test should focus on the use case, so common validation might not even be required.

Reply:
Good points!
Addressed all 3. Speaking of the last (focus) - I also prefer when fail-tests are not duplicating the validation.